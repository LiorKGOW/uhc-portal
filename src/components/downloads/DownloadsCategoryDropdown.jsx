import React from 'react';
import PropTypes from 'prop-types';
import {
  FormSelect,
  FormSelectOption,
} from '@patternfly/react-core';

export const downloadsCategoryTitles = {
  ALL: 'All categories',
  CLI: 'Command-line interface (CLI) tools',
  DEV: 'Developer tools',
  INSTALLATION: 'OpenShift installation',
  TOKENS: 'Tokens',
};

const DownloadsCategoryDropdown = ({ selectedCategory, setCategory }) => (
  <FormSelect
    aria-label="Select category"
    value={selectedCategory}
    onChange={setCategory}
  >
    {['ALL', 'CLI', 'DEV', 'INSTALLATION', 'TOKENS'].map(key => (
      <FormSelectOption key={key} value={key} label={downloadsCategoryTitles[key]} />
    ))}
  </FormSelect>
);
DownloadsCategoryDropdown.propTypes = {
  selectedCategory: PropTypes.oneOf(Object.keys(downloadsCategoryTitles)).isRequired,
  setCategory: PropTypes.func.isRequired,
};

export default DownloadsCategoryDropdown;
