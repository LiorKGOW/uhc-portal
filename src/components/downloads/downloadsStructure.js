import { tools } from '../../common/installLinks.mjs';

/** Used to track row collapsed/expanded state, and also for URLs linking to specific row. */
export const expandKeys = {
  ...tools,
  // the rest should be distinct from `tools` keys.
  PULL_SECRET: 'pull-secret',
  TOKEN_OCM: 'ocm-api-token',
};

/**
 * Represents the sections and entries of downloads page.
 */
export const downloadsCategories = [
  // 'ALL' will be inserted here, see below.
  {
    key: 'CLI',
    title: 'Command-line interface (CLI) tools',
    tools: [tools.OC, tools.OCM, tools.ROSA, tools.KN, tools.TKN],
  },
  {
    key: 'DEV',
    title: 'Developer tools',
    tools: [tools.ODO, tools.HELM, tools.OPM, tools.OPERATOR_SDK, tools.RHOAS],
  },
  {
    key: 'INSTALLATION',
    title: 'OpenShift installation',
    tools: [
      tools.X86INSTALLER,
      tools.IBMZINSTALLER,
      tools.PPCINSTALLER,
      tools.ARMINSTALLER,
      tools.MULTIINSTALLER,
      tools.CRC,
    ],
  },
  {
    key: 'DISCONNECTED_INSTALLATION',
    title: 'OpenShift disconnected installation tools',
    tools: [tools.MIRROR_REGISTRY, tools.OC_MIRROR_PLUGIN],
  },
  {
    key: 'CUSTOM_INSTALLATION',
    title: 'OpenShift installation customization tools',
    tools: [tools.BUTANE, tools.COREOS_INSTALLER, tools.CCOCTL],
  },
  {
    key: 'TOKENS',
    title: 'Tokens',
    tools: [expandKeys.PULL_SECRET, expandKeys.TOKEN_OCM],
  },
];
downloadsCategories.unshift({
  key: 'ALL',
  title: 'All categories',
  tools: [].concat(...downloadsCategories.map((c) => c.tools)),
});
