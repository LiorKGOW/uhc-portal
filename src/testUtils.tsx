import React from 'react';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import { render, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { toHaveNoViolations, axe } from 'jest-axe';

import { createBrowserHistory } from 'history';
import { store } from './redux/store';
import { reduxReducers } from './redux/reducers';
import * as restrictedEnv from './restrictedEnv';

const history = createBrowserHistory();

export const TestWrapper = ({
  children,
  initialStore,
}: {
  children: React.ReactNode;
  initialStore?: Store;
}) => <Provider store={initialStore || store}>{children}</Provider>;

const renderWithState = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialState?: any, // setting to any because only a partial state structure may be sent
) => {
  const newStore = initialState ? createStore(reduxReducers(history), initialState) : store;

  const Wrapper = ({ children }: { children: React.ReactNode }) => (
    <TestWrapper initialStore={newStore}>{children}</TestWrapper>
  );
  return { ...render(ui, { wrapper: Wrapper, ...options }), user: userEvent.setup() };
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// Type not exported in the library
export type UserEventType = {
  type: (
    field: HTMLElement,
    typeValue: string,
    { initialSelectionStart }: { initialSelectionStart: number },
  ) => void;
  clear: (field: HTMLElement) => void;
};

export { renderWithState as render };

/* ***** Items outside of React Test Library ************ */
expect.extend(toHaveNoViolations);

export const checkAccessibility = async (container: HTMLElement | string) => {
  const results = await axe(container);
  expect(results).toHaveNoViolations();
};

export const insightsMock = () => {
  global.insights = {
    chrome: {
      ...global.insights.chrome,
      on: () => () => {},
      appNavClick: () => {},
      auth: {
        getUser: () => Promise.resolve({ data: {} }),
        getToken: () => Promise.resolve(),
        getOfflineToken: () => Promise.resolve({ data: { refresh_token: 'hello' } }),
      },
    },
  };
};

export const mockRestrictedEnv = () => {
  const mock = jest.spyOn(restrictedEnv, 'isRestrictedEnv');
  mock.mockReturnValue(false);
  return mock;
};
