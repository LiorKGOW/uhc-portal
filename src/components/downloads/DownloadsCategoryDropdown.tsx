import React from 'react';

import { FormSelect, FormSelectOption } from '@patternfly/react-core';

import { allCategories, downloadsCategories } from './downloadsStructure';

type DownloadsCategoryDropdownProps = {
  selectedCategory: (typeof allCategories)[number]['key'];
  setCategory: (
    event: React.FormEvent<HTMLSelectElement>,
    selectedCategory: (typeof allCategories)[number]['key'],
  ) => void;
};

const DownloadsCategoryDropdown = ({
  selectedCategory,
  setCategory,
}: DownloadsCategoryDropdownProps) => (
  <FormSelect
    aria-label="Select category"
    value={selectedCategory}
    onChange={setCategory}
    data-testid="downloads-category-dropdown"
  >
    {downloadsCategories().map((c) => (
      <FormSelectOption key={c.key} value={c.key} label={c.title} />
    ))}
  </FormSelect>
);

export default DownloadsCategoryDropdown;
