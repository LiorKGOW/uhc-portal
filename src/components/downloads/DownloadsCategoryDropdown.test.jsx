import React from 'react';

import { checkAccessibility, render, screen } from '~/testUtils';

import DownloadsCategoryDropdown from './DownloadsCategoryDropdown';
import { downloadsCategories } from './downloadsStructure';

describe('<DownloadsCategoryDropdown />', () => {
  const setCategory = jest.fn();
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('is accessible', async () => {
    const { container } = render(
      <DownloadsCategoryDropdown selectedCategory="DEV" setCategory={setCategory} />,
    );
    await checkAccessibility(container);
  });

  it('has expected number of options', () => {
    render(<DownloadsCategoryDropdown selectedCategory="DEV" setCategory={setCategory} />);

    expect(screen.getAllByRole('option')).toHaveLength(downloadsCategories().length);
  });
});
