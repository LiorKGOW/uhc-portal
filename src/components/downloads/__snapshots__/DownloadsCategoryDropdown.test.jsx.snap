// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadsCategoryDropdown /> should render 1`] = `
<FormSelect
  aria-label="select-category"
  className=""
  isDisabled={false}
  isRequired={false}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  ouiaSafe={true}
  validated="default"
  value="DEV"
>
  <FormSelectOption
    key="ALL"
    label="All categories"
    value="ALL"
  />
  <FormSelectOption
    key="CLI"
    label="Command-line interface (CLI) tools"
    value="CLI"
  />
  <FormSelectOption
    key="DEV"
    label="Developer tools"
    value="DEV"
  />
  <FormSelectOption
    key="INSTALLATION"
    label="OpenShift installation"
    value="INSTALLATION"
  />
  <FormSelectOption
    key="TOKENS"
    label="Tokens"
    value="TOKENS"
  />
</FormSelect>
`;
